### PR TITLE
fix to MSBuild path not found error

### PR DESCRIPTION
**Pull request by mrkdeng**
## Description
Two changes to address this issue:
1. Log error without throwing then catching an error
2. Refactored methods to make it more readable 

## Additional context
---
*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs [hunk 1]
```diff
@@ -560,36 +560,40 @@ private EnvironmentOptions GetEnvironmentOptions(bool requiresNetFramework, stri
...
+                        }
                     }
-                    if (!string.IsNullOrEmpty(msbuildExe)) options.EnvironmentVariables.Add(EnvironmentVariables.MSBUILD_EXE_PATH, msbuildExe);
-                    else { throw new Exception(); }
-
-                }
-                catch (Exception ex)
-                {
-                    Logger.LogError(ex, "Build error: Codelyzer wasn't able to retrieve the MSBuild path");
+                    catch (NullReferenceException ex)
```
Comment (jonlouie) [inline src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:587]: Suggestions: 
* Throw an Exception instead of NullReferenceException and keep the catch statement as `catch (Exception ex)` so we can catch any other unexpected errors.
* Add an error message to the exception constructor

File: src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs [hunk 2]
```diff
@@ -560,36 +560,40 @@ private EnvironmentOptions GetEnvironmentOptions(bool requiresNetFramework, stri
...
+                    {
+                        Logger.LogError(ex, "Build error: Codelyzer wasn't able to retrieve the MSBuild path. " +
+                            "Visual Studio and MSBuild might not be installed.");
```
Comment (jonlouie) [inline src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:590]: We should be able to just log the exception object here if the error message is added directly to the exception. Also, remove the "Build error: " prefix to prevent the error message from being categorized incorrectly.

**Review by jonlouie — CHANGES_REQUESTED** [review]

File: src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs (hunk 1 above)
Comment (ppittle) [reply src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:587]: To expand on this - you should not throw an exception just to immediately catch it - exception handling isn't designed to be used for normal program branching as it's very expensive to throw an exception.  Instead try:

```csharp
if (os == OSPlatform.Linux || os == OSPlatform.OSX)
{
    options.EnvironmentVariables.Add(EnvironmentVariables.MSBUILD_EXE_PATH, Constants.MsBuildCommandName);
}
else
{
    var msbuildExe = _analyzerConfiguration.BuildSettings.MSBuildPath;
    
    if (string.IsNullOrEmpty(msbuildExe))
       msbuildExe = _msBuildDetector.GetFirstMatchingMsBuildFromPath(toolsVersion: toolsVersion);
       
    if (!string.IsNullOrEmpty(msbuildExe))
    {
             options.EnvironmentVariables.Add(EnvironmentVariables.MSBUILD_EXE_PATH, msbuildExe);
      }
      else 
      { 
          Logger.LogError("Build error: Codelyzer wasn't able to retrieve the MSBuild path. " +
                            "Visual Studio and MSBuild might not be installed.");
      }
}
```

File: src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs [hunk 3]
```diff
@@ -560,36 +560,40 @@ private EnvironmentOptions GetEnvironmentOptions(bool requiresNetFramework, stri
...
-                {
-                    options.EnvironmentVariables.Add(EnvironmentVariables.MSBUILD_EXE_PATH, Constants.MsBuildCommandName);
-                }
-            }
-
             //We want to provide the MsBuild path only if it's a framework solution. Buildalyzer automatically builds core solutions using "dotnet"
             if (requiresNetFramework)
             {
-                try
+                if (os == OSPlatform.Linux || os == OSPlatform.OSX)
```
Comment (ppittle) [inline src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:566]: Why did you move the `if (os == OSPlatform.Linux || os == OSPlatform.OSX)` inside the `if (requiresNetFramework)`?

```
 //We want to provide the MsBuild path only if it's a framework solution. Buildalyzer automatically builds core solutions using "dotnet"
if (requiresNetFramework)
{
    if (os == OSPlatform.Linux || os == OSPlatform.OSX)
```

Is it a valid use case that the OS would be linux if the solution is NetFramework as Framework only runs on Windows?

Comment (mrkdeng) [reply src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:566]: @ppittle @jonlouie 
I was following the old logic
```
            if (os == OSPlatform.Linux || os == OSPlatform.OSX)
            {
                if (requiresNetFramework)
                {
                    options.EnvironmentVariables.Add(EnvironmentVariables.MSBUILD_EXE_PATH, Constants.MsBuildCommandName);
                }
            }
```
If `(os == OSPlatform.Linux || os == OSPlatform.OSX)` and `requiresNetFramework` are mutually exclusive, I doubt we ever hit this point.

Comment (jonlouie) [reply src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs]: I think we have to be careful about this case. It's true that unix platforms won't build .NET Framework projects, but nothing is stopping developers from trying to analyze a .NET Framework project from a linux/mac machine. Also, look at like 296 of [this diff](https://github.com/aws/codelyzer/commit/97272f24821075a57f82505d14e698c71d690ec4#diff-3ea344cda7ecb78aa25b568af9c5c7024c50d0c6f6ad859dc2f4b779a7bb73b9). The `requiresNetFramework` check was added here intentionally. Unless we fully understand the consequences of removing it, we should let it remain.

To make the intent of this code more clear, I would suggest refactoring the code into a separate function similar to this:
```csharp
if (requiresNetFramework)
{
    SetMSBuildPathEnvironmentVariable(options);
}
.
.
.
private void SetMSBuildPathEnvironmentVariable(EnvironmentOptions options)
{
    var os = DetermineOSPlatform();
    if (os == OSPlatform.Linux || os == OSPlatform.OSX)
    {
        options.EnvironmentVariables.Add(EnvironmentVariables.MSBUILD_EXE_PATH, Constants.MsBuildCommandName);
    }
    else  // os == OSPlatform.Windows
    {
        var msbuildExe = _analyzerConfiguration.BuildSettings.MSBuildPath;
        if (string.IsNullOrEmpty(msbuildExe))
        {
            msbuildExe = _msBuildDetector.GetFirstMatchingMsBuildFromPath(toolsVersion: toolsVersion);
        }
        if (!string.IsNullOrEmpty(msbuildExe)) 
        {
            options.EnvironmentVariables.Add(EnvironmentVariables.MSBUILD_EXE_PATH, msbuildExe);
        }
        else 
        {
            Logger.LogError("Codelyzer wasn't able to retrieve the MSBuild path. " +
                "Visual Studio and MSBuild might not be installed.");
        } 
    }
}
```

Comment (mrkdeng) [reply src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs]: I agree with not removing the `requiresNetFramework` check. To your second point - I didn't use this pattern often, in this case it tries to set for `EnvironmentOptions` depending on an outsider parameter `DetermineOSPlatform `. But I see this is how we avoid using `try/catch`. Is there a practice that we should follow regarding the codebase?

**Review by jonlouie — APPROVED** [review]

**Review by saikiranakula-amzn — APPROVED** [review]
